tensorflow/python/kernel_tests/math_ops/reduction_ops_test.py

```diff
@@ -310,6 +310,52 @@ def testFloat32(self):
             tf_out_sum_xz, tf_out_sum_y = self.evaluate([tf_sum_xz, tf_sum_y])
           self.assertAllClose(sum_y, tf_out_sum_y)
           self.assertAllClose(sum_xz, tf_out_sum_xz)
+  
+  @test_util.run_deprecated_v1
+  def testFloat32BFloat16(self):
+    for dtype in [dtypes.float32, dtypes.bfloat16]:
+      dtype_np = np.float32 if dtype == dtypes.float32 else dtypes.bfloat16.as_numpy_dtype
+      for rank in range(1, _MAX_RANK + 1):
+        np_arr = self._makeIncremental((2,) * rank, dtype)
+        self._compareAllAxes(np_arr)
+
+      for _ in range(10):
+        size_x = int(2**np.random.uniform(0, 15))
+        size_y = int(2**np.random.uniform(0, 15))
+
+        if size_x * size_y > 1e7:
+          size_y = int(1e7 / size_x)
+
+        arr = np.ones([size_x, size_y], dtype=dtype_np)
+        col_sum = np.sum(arr, axis=0, dtype=np.float32)
+        row_sum = np.sum(arr, axis=1, dtype=np.float32)
+
+        with self.session(graph=ops.Graph(), use_gpu=True) as sess:
+          tf_row_sum = self._tf_reduce(arr, 1, False)
+          tf_col_sum = self._tf_reduce(arr, 0, False)
+          tf_out_row, tf_out_col = self.evaluate([tf_row_sum, tf_col_sum])
+        if dtype == dtypes.bfloat16:
```

Comment:
Why do we need the conversion to the numpy type?

Reply:
In the test, to verify the output of tf_reduce sum in BF16 we compare it with np.sum(). But np.sum() seems to have issues with bfloat16. I will illustrate the issue with an example.
A **arr=np.ones([512, 512], dtype=dtypes.bfloat16.as_numpy_dtype)**, this will initialize a numpy ndarray of size **512x512** will all ones. Further, if **np.sum(arr, axis=1)** is performed, the expected output should be numpy array with size=512 and sum[0],...sum[511] = 512 (i.e. all the elements in summation equals 512). But with BFloat16 as dtype np.sum() returns array of size=512 with all elements equal to 256. This is only observed with array of larger size. Interestingly, if an ndarray of 512x512 size has all 2s, the sum across the rows returns an array of size 512, with all elements equal to 512 instead of 1024. Which suggests np.sum() with BF16 only considers first 256 elements from the array.

The following merged PR: https://github.com/tensorflow/tensorflow/pull/57989, also fails when AddN is run with an input of larger size. This issue seems to be from numpy end only as TF sum works perfectly fine with inputs of large size with BF16.

Reply:
It's because bfloat16, which only has 7 bits of precision, can't represent the number 257, so `256+1` gets rounded down to 256, and you get stuck here.  Adding by 2s, you skip over that one issue, but then run into where bfloat16 can't represent 513, 514 or 515, so `512+2` is rounded down to 512 and you get stuck again.  Numpy seems to be doing a naive sequential sum here.

You should probably modify your inputs so you're not summing a bunch of ones.

Reply:
I have updated the input size by bringing it below 256. I tried working with random decimal values, but due to the difference in precision handling for TF and Numpy, the test was still failing. As this being a numpy issue, bringing the input size down to avoid the rounding down error, passes the test.